crates/wasmtime/src/runtime/vm/component/libcalls.rs

```diff
@@ -871,7 +873,18 @@ unsafe fn error_context_transfer(
 }
 
 #[cfg(feature = "component-model-async")]
-unsafe fn future_new(vmctx: NonNull<VMComponentContext>, ty: u32) -> Result<u32> {
+unsafe impl HostResultHasUnwindSentinel for ResourcePair {
+    type Abi = u64;
+    const SENTINEL: u64 = u64::MAX;
+
+    fn into_abi(self) -> Self::Abi {
+        assert!(self.read & (1 << 31) == 0);
+        (u64::from(self.read) << 32) | u64::from(self.write)
```

Comment:
To confirm the high bit of this handle is always 0, right? (double-checking my bit-packing here)

FWIW I realized I'd rather quite avoid taking a pointer on the intrinsic since then you need to attach a memory to it with canonical options and all and that's a bit of a pain.

Reply:
The high two bits are unset, yes.  I think the plan was to be even more aggressive and reserve the high four bits per [`MAX_LENGTH = 2**28 - 1` in this section](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#table-state), but definitely at least two.